data/lib/src/datasource_impl/contact_datasource_impl.dart

```diff
@@ -0,0 +1,57 @@
+// LinShare is an open source filesharing software, part of the LinPKI software
+// suite, developed by Linagora.
+//
+// Copyright (C) 2021 LINAGORA
+//
+// This program is free software: you can redistribute it and/or modify it under the
+// terms of the GNU Affero General Public License as published by the Free Software
+// Foundation, either version 3 of the License, or (at your option) any later version,
+// provided you comply with the Additional Terms applicable for LinShare software by
+// Linagora pursuant to Section 7 of the GNU Affero General Public License,
+// subsections (b), (c), and (e), pursuant to which you must notably (i) retain the
+// display in the interface of the “LinShare™” trademark/logo, the "Libre & Free" mention,
+// the words “You are using the Free and Open Source version of LinShare™, powered by
+// Linagora © 2009–2021. Contribute to Linshare R&D by subscribing to an Enterprise
+// offer!”. You must also retain the latter notice in all asynchronous messages such as
+// e-mails sent with the Program, (ii) retain all hypertext links between LinShare and
+// http://www.linshare.org, between linagora.com and Linagora, and (iii) refrain from
+// infringing Linagora intellectual property rights over its trademarks and commercial
+// brands. Other Additional Terms apply, see
+// <http://www.linshare.org/licenses/LinShare-License_AfferoGPL-v3.pdf>
+// for more details.
+// This program is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+// FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for
+// more details.
+// You should have received a copy of the GNU Affero General Public License and its
+// applicable Additional Terms for LinShare along with this program. If not, see
+// <http://www.gnu.org/licenses/> for the GNU Affero General Public License version
+//  3 and <http://www.linshare.org/licenses/LinShare-License_AfferoGPL-v3.pdf> for
+//  the Additional Terms applicable to LinShare software.
+
+import 'package:contacts_service/contacts_service.dart';
+import 'package:data/src/datasource/contact_datasource.dart';
+import 'package:domain/domain.dart';
+import 'package:domain/src/model/autocomplete/autocomplete_pattern.dart';
+import 'package:domain/src/model/contact/contact.dart' as contact;
+
+class ContactDataSourceImpl extends ContactDataSource {
+
+  @override
+  Future<List<contact.Contact>> getContactSuggestions(AutoCompletePattern autoCompletePattern) async {
+    final suggestedList = await ContactsService.getContactsByEmailOrName(autoCompletePattern.value);
+    final deviceContactList = <contact.Contact>[];
+    suggestedList.forEach((suggestion) {
```

Comment:
WDYT, if using map.

Reply:
oh you mean using map here to return a different type directly? I don't think that will work in my case. I also need to check if the contact return from the library have more than 1 email, then I have to split that into 2 different `DeviceContact`

Reply:
`flatMap` then convert to deviceContact. I don't like to add each item to a shared list.

Reply:
thank guys, I think its better now.